### PR TITLE
File Watcher Authorization Server Interceptor

### PR DESCRIPTION
**Pull request by ashithasantosh**
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc-java/9775)

### CODE REVIEW

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+ * Authorization server interceptor for policy from file with refresh capability.
+ * The class will get <a href="https://github.com/grpc/proposal/blob/master/A43-grpc-authorization-api.md#user-facing-authorization-policy">
+ * gRPC Authorization policy</a> from a JSON file during initialization.
+ */
+public final class FileWatcherAuthorizationServerInterceptor implements ServerInterceptor {
+  private static final Logger logger = 
+      Logger.getLogger(FileWatcherAuthorizationServerInterceptor.class.getName());
+
+  private volatile AuthorizationServerInterceptor internalAuthzServerInterceptor;
+  private final ScheduledExecutorService scheduledExecutorService =
```
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java:50]: Have the user pass this in to `scheduleRefreshes()`. For reference, see `AdvancedTlsX509TrustManager.updateTrustCredentialsFromFile()`.

Then, in the tests pass FakeClock. That will then allow you to modify the policy file, forwardTime() (commonly we do this in two phases, forward time too little, check the behavior is still the old behavior, then forward time the small remaining amount), and then check the interceptor behavior changed. You wouldn't need `testCallback`.

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                    .setDaemon(true)
+                    .build());
+
+  private final File policyFile;
+  private FileTime lastModifiedTime;
+
+  private Thread testCallback;
+
+  private FileWatcherAuthorizationServerInterceptor(File policyFile)
+        throws IllegalArgumentException, IOException {
```
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java:63]: `IllegalArgumentException` is a `RuntimeException`, so doesn't need to be listed.

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    final ScheduledFuture<?> future = 
+        scheduledExecutorService.scheduleWithFixedDelay(new Runnable() {
+          @Override public void run() {
+            try {
+                updateInternalInterceptor();
+            } catch (Exception e) {
+                logger.log(Level.WARNING, "Authorization Policy file reload failed: " + e);
+            }
+            }
```
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java:111]: Unindent.

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    if (period <= 0) {
+      throw new IllegalArgumentException("Refresh interval must be greater than 0");
+    }
+    final ScheduledFuture<?> future = 
+        scheduledExecutorService.scheduleWithFixedDelay(new Runnable() {
+          @Override public void run() {
+            try {
+                updateInternalInterceptor();
+            } catch (Exception e) {
+                logger.log(Level.WARNING, "Authorization Policy file reload failed: " + e);
```
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java:109]: Pass `e` as its own argument.

https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#log-java.util.logging.Level-java.lang.String-java.lang.Throwable-

File: authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class FileWatcherAuthorizationServerInterceptorTest {
+  @Test
+  public void invalidPolicyFailsAuthzInterceptorCreation() throws Exception {
+    File policyFile = File.createTempFile("temp", "json");
+    policyFile.deleteOnExit();
+    try (FileOutputStream outputStream = new FileOutputStream(policyFile)) {
```
Comment (ejona86) [inline authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java:38]: Feel free to use Guava's [`Files.write(byte[], File)`](https://guava.dev/releases/snapshot/api/docs/com/google/common/io/Files.html#write(byte%5B%5D,java.io.File)) and/or create a helper function to help make the test more terse.

**Review by ashithasantosh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/grpc/grpc-java/9775)* status: 0 of 5 files reviewed, 13 unresolved discussions (waiting on @ejona86)

*[a discussion](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdipu77HFL6l-0ZQob:-NPdipu77HFL6l-0ZQoc:b44zsg4) (no related file):*
Thank you for the review:) Please review the updated code.

___
___
*[`authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java` line 50 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm3-r1-50:-NPdZwlnBumyXqRxvdHX:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java#L50)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Have the user pass this in to `scheduleRefreshes()`. For reference, see `AdvancedTlsX509TrustManager.updateTrustCredentialsFromFile()`.

Then, in the tests pass FakeClock. That will then allow you to modify the policy file, forwardTime() (commonly we do this in two phases, forward time too little, check the behavior is still the old behavior, then forward time the small remaining amount), and then check the interceptor behavior changed. You wouldn't need `testCallback`.
</blockquote></details>

Done.
___
*[`authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java` line 63 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm3-r1-63:-NPdZxvg3Cxd09MO_nIY:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java#L63)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

`IllegalArgumentException` is a `RuntimeException`, so doesn't need to be listed.
</blockquote></details>

Done.
___
*[`authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java` line 99 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm3-r1-99:-NPd_5805ZRjn-fwGXGe:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java#L99)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Intent two more spaces, since it is a line continuation. Ditto for create().

(Or don't wrap the line)
</blockquote></details>

Done.
___
*[`authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java` line 109 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm3-r1-109:-NPd_9YW4UaRbzmQG9wN:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java#L109)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Pass `e` as its own argument.

https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html#log-java.util.logging.Level-java.lang.String-java.lang.Throwable-
</blockquote></details>

Done.
___
*[`authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java` line 111 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm3-r1-111:-NPd_BXLD3kTyGCJzHV2:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java#L111)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Unindent.
</blockquote></details>

Done.
___
*[`authz/src/test/java/io/grpc/authz/AuthorizationEnd2EndTest.java` line 447 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm2-r1-447:-NPd_MCJ09zYxEBaoyg0:bsiumu9) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/AuthorizationEnd2EndTest.java#L447)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

This is a strange use of `deleteOnExit()`, because it seems it could just be `delete()` instead. I'd expect any `deleteOnExit()` to be just after the file is created, so that if an exception is thrown later in the test the file would still get deleted.
</blockquote></details>

Moved the logic to delete policy file to tearDown()
___
*[`authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java` line 36 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm4-r1-36:-NPdiWeh8BNznUL7dY5h:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java#L36)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Sometimes it is convenient to use `createTempFile()`+`deleteOnExit()` in tests, although [`TemporaryFolder`](https://junit.org/junit4/javadoc/4.13/org/junit/rules/TemporaryFolder.html) is generally preferred because it will be cleaned up at the end of the test instead of if/when the JVM exits cleanly.

Because this is an easy case, you could also do:
```java
private File policyFile = File.createTempFile("temp", "json");

@After public void cleanUp() {
  policyFile.delete(); // does not throw if file not found
}
```
</blockquote></details>

Done.
___
*[`authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java` line 38 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm4-r1-38:-NPd_zf13yhOUUPhtb6W:bkh7u3t) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java#L38)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Feel free to use Guava's [`Files.write(byte[], File)`](https://guava.dev/releases/snapshot/api/docs/com/google/common/io/Files.html#write(byte%5B%5D,java.io.File)) and/or create a helper function to help make the test more terse.
</blockquote></details>

Thank you for the suggestion. Updated the tests to use Guava API.
___
*[`authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java` line 41 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm4-r1-41:-NPdaG_S2dKr-stZKTpG:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java#L41)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

The try-with-resources will close the output stream, so this line can be deleted.
</blockquote></details>

Done.
___
*[`authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java` line 48 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm4-r1-48:-NPdaIGY6zlWJHl1VgtU:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java#L48)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Do not assert messages that you don't generate. You could maybe check for "malformed," but this full message is far too precise and would trivially break on an upgrade to GSON.
</blockquote></details>

Done.
___
*[`authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java` line 50 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm4-r1-50:-NPdaJqLCF9_X8vm7NR5:b-896fix) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java#L50)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

Delete this catch and let the exception propagate.
</blockquote></details>

Done.
___
*[`authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java` line 84 at r1](https://reviewable.io/reviews/grpc/grpc-java/9775#-NPdT8u8BEvNUZ0C_tm4-r1-84:-NPdiLmP5E_Daj8pxY1B:bk81ked) ([raw file](https://github.com/grpc/grpc-java/blob/6f7fef0a2958b7f3ea92f9560987fcba9e98bd40/authz/src/test/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptorTest.java#L84)):*
<details><summary><i>Previously, ejona86 (Eric Anderson) wrote…</i></summary><blockquote>

What is the point of this line? It is leaking the returned Closeable.
</blockquote></details>

Deleted the line. I agree it does not make sense here.

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java (hunk 1 above)
Comment (ashithasantosh) [reply authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java]: Sounds good. Updated the code to use FakeClock.

Thank you for the review! I have all the comments. @ejona86 PTAL

**Review by larry-safran — APPROVED** [review]
There is one stylistic review comment not addressed, which is about not relying on deleteOnExit for a test to cleanup the temporary file.

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java [hunk 6]
```diff
@@ -93,22 +76,25 @@ void updateInternalInterceptor() throws IOException {
...
+        executor.scheduleWithFixedDelay(new Runnable() {
+          @Override
+          public void run() {
             try {
-                updateInternalInterceptor();
+              updateInternalInterceptor();
             } catch (Exception e) {
-                logger.log(Level.WARNING, "Authorization Policy file reload failed: " + e);
-            }
+              logger.log(Level.WARNING, "Authorization Policy file reload failed: ", e);
```
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java:95]: Remove ": " from string. `e` is not a formatted argument. It is its own separate thing, displayed however the logging framework wants.

File: authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java [hunk 7]
```diff
@@ -47,20 +43,11 @@ public final class FileWatcherAuthorizationServerInterceptor implements ServerIn
...
   private volatile AuthorizationServerInterceptor internalAuthzServerInterceptor;
-  private final ScheduledExecutorService scheduledExecutorService = 
-      Executors.newSingleThreadScheduledExecutor(
-                    new ThreadFactoryBuilder()
-                    .setNameFormat("filewatcher" + "-%d")
-                    .setDaemon(true)
-                    .build());
 
   private final File policyFile;
-  private FileTime lastModifiedTime;
```
Comment (ejona86) [inline authz/src/main/java/io/grpc/authz/FileWatcherAuthorizationServerInterceptor.java]: FYI, for testing you can do something like:
https://github.com/grpc/grpc-java/blob/1551cc7403e0f9bcb3b949be8c486ce207395c6b/xds/src/test/java/io/grpc/xds/internal/security/certprovider/FileWatcherCertificateProviderTest.java#L117-L118

**Review by ejona86 — APPROVED** [review]
We should probably fix the formatting of the one string before merging.

**Pull request merged**